From 756fbaa51410943f42deab829e1c96487e1014ed Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Sat, 25 Oct 2025 16:18:03 +0200
Subject: [PATCH 0992/1013] evl/net: ipv4: fix probing of connected/stale
 neighbour on solicit

Upon solicitation, we might find a connected neighbour into the inband
ARP table which is not present in our front cache, or a stale
neighbour entry, which would yield the following issues:

- for an entry in connected state but not present in our front cache
(*), the inband ARP logic would not send a probe request due to the
state, which would in turn prevent the peer to be added to our front
cache since it listens for successful probes from the inband ARP
management.

- for an entry in stale state - therefore absent from our front cache
by definition - the inband ARP logic would not send a probe request
either.

Either way, this would cause evl_net_ipv4_solicit() to bail out with a
timeout error.

To fix this, force update our front cache with any connected neighbour
received from the routing process, instead of asking the inband ARP
management to probe for the peer. In the same move, make sure to
invalidate/expire stale entries to force probing instead.

(*) typically, this may happen if the front cache is forcibly flushed
but the inband ARP table is not.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 include/evl/net/ipv4/arp.h |  2 ++
 kernel/evl/net/ipv4/arp.c  | 26 ++++++++++++++++++-
 kernel/evl/net/ipv4/ipv4.c | 52 +++++++++++++++++++++++++++++++-------
 3 files changed, 70 insertions(+), 10 deletions(-)

diff --git a/include/evl/net/ipv4/arp.h b/include/evl/net/ipv4/arp.h
index f6715ce7e989..07f01aa75958 100644
--- a/include/evl/net/ipv4/arp.h
+++ b/include/evl/net/ipv4/arp.h
@@ -34,6 +34,8 @@ void evl_net_cleanup_arp(struct net *net);
 
 void evl_net_flush_arp(struct net *net);
 
+int evl_net_update_arp(struct neighbour *neigh);
+
 struct evl_net_arp_entry *
 evl_net_get_arp_entry(struct net_device *dev, __be32 addr);
 
diff --git a/kernel/evl/net/ipv4/arp.c b/kernel/evl/net/ipv4/arp.c
index b9ad6a871efe..296426e50a01 100644
--- a/kernel/evl/net/ipv4/arp.c
+++ b/kernel/evl/net/ipv4/arp.c
@@ -99,7 +99,7 @@ static struct evl_cache_ops arp_cache_ops = {
 };
 
 /*
- * Cache a new ARP entry.
+ * Cache a new ARP entry. A previous match is replaced by the new one.
  */
 static int cache_arp_entry(struct evl_cache *cache, struct neighbour *neigh) /* in-band */
 {
@@ -209,6 +209,30 @@ struct evl_net_arp_entry *evl_net_get_arp_entry(struct net_device *dev, __be32 a
 	return NULL;
 }
 
+/*
+ * Add a new neighbour to our ARP front cache. Any previous match is
+ * replaced by the new one.
+ */
+int evl_net_update_arp(struct neighbour *neigh) /* inband */
+{
+	struct oob_net_state *nets = &dev_net(neigh->dev)->oob;
+	struct evl_cache *cache = &nets->ipv4.arp;
+	int ret = -ESTALE;
+
+	read_lock_bh(&neigh->lock);
+
+	/*
+	 * We only cache entries for connected neighbours: recheck
+	 * under lock.
+	 */
+	if (READ_ONCE(neigh->nud_state) & NUD_CONNECTED)
+		ret = cache_arp_entry(cache, neigh);
+
+	read_unlock_bh(&neigh->lock);
+
+	return ret;
+}
+
 static struct evl_net_arp_entry *
 fill_pseudo_arp(struct net_device *dev, __be32 ipaddr,
 		const u8 *hwaddr, int hwlen,
diff --git a/kernel/evl/net/ipv4/ipv4.c b/kernel/evl/net/ipv4/ipv4.c
index 6ce31d855f14..d32be2a7fc73 100644
--- a/kernel/evl/net/ipv4/ipv4.c
+++ b/kernel/evl/net/ipv4/ipv4.c
@@ -231,6 +231,42 @@ void __evl_net_ipv4_gc(struct evl_net_frag_tdir *ftdir)
 	evl_unlock_kmutex(&ftdir->lock);
 }
 
+/*
+ * Check whether a connected neighbour is available from our front
+ * cache, otherwise solicit the peer who should respond to the ARP
+ * request within the allotted time. In the latter case, probing is
+ * forced for stale entries.
+ */
+static int check_probe_neighbour(struct neighbour *neigh)
+{
+	u8 nud_state = READ_ONCE(neigh->nud_state);
+	int ret;
+
+	if (EVL_WARN_ON(NET, nud_state & NUD_NOARP))
+		return -EINVAL;
+
+	if (nud_state & NUD_CONNECTED) {
+		ret = evl_net_update_arp(neigh);
+		if (ret != -ESTALE)
+			return ret;
+		nud_state = READ_ONCE(neigh->nud_state);
+	}
+
+	if (nud_state & NUD_STALE) {
+		/* Invalidate/expire a stale entry to force probing. */
+		ret = neigh_update(neigh, NULL, NUD_NONE,
+				NEIGH_UPDATE_F_OVERRIDE|
+				NEIGH_UPDATE_F_ADMIN, 0);
+		if (ret)
+			return ret;
+	}
+
+	/* Start the probing process. */
+	neigh_event_send(neigh, NULL);
+
+	return 0;
+}
+
 /**
  * evl_net_ipv4_solicit - Resolve an IPv4 address into a link-layer
  * address using ARP neighbour solicitation. This call waits for the
@@ -249,7 +285,7 @@ void __evl_net_ipv4_gc(struct evl_net_frag_tdir *ftdir)
  */
 int evl_net_ipv4_solicit(struct net *net,
 			struct net_device *dev,
-			struct sockaddr *addr, int flags)
+			struct sockaddr *addr, int flags) /* inband */
 {
 	struct evl_net_arp_entry *e = NULL;
 	struct neighbour *neigh;
@@ -295,18 +331,16 @@ int evl_net_ipv4_solicit(struct net *net,
 	}
 
 	if (likely(!(neigh->nud_state & NUD_NOARP))) {
-		/*
-		 * Solicit the peer which should respond to ARP
-		 * requests within the allotted time.
-		 */
-		neigh_event_send(neigh, NULL);
+		ret = check_probe_neighbour(neigh);
+		if (ret)
+			goto out;
 		ret = wait_event_interruptible_timeout(evl_arp_event,
-			       (e = evl_net_get_arp_entry(dev, ipaddr)),
-				evl_net_ipv4_solicit_timeout * HZ
+						(e = evl_net_get_arp_entry(dev, ipaddr)),
+						evl_net_ipv4_solicit_timeout * HZ
 			);
 		ret = e ? 0 : ret ?: -ETIMEDOUT;
 	}
-
+out:
 	evl_net_put_dev(dev);
 
 	if (e) {
-- 
2.30.2

